[nikohomecontrol] Energy access

mherwege opened the pull request:
Depends on [PR #12885](https://github.com/openhab/openhab-addons/pull/12885)

This PR adds:
- NHC I energy meters
- NHC II access control

To reduce code duplication with the added functionality, it also contains code refactoring to lift common logic to base classes. This makes the changes quite substantial. I am no longer able to split it in smaller pieces, as there are too many dependencies on previous PR's only just reviewed or still in review, which makes rebasing a nightmare. I was able to rebase this on  [PR #12885](https://github.com/openhab/openhab-addons/pull/12885), but had to squash the detailed commits to make it feasible.

This PR complements [PR #11963](https://github.com/openhab/openhab-addons/pull/11963), improving thermostats, starting from the same base PR. After review and merge of  [PR #12885](https://github.com/openhab/openhab-addons/pull/12885) I will rebase both PR's.

mherwege commented:
@lolodomo rebased as well.

lolodomo commented:
There us no hurry but a rebase of your branch will be required to solve current conflicted files.

mherwege commented:
Rebased

lolodomo commented on bundles/org.openhab.binding.nikohomecontrol/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -186,20 +179,159 @@
+			<channel id="energylast" typeId="measurementtime"/>
+		</channels>
+		<config-description>
+			<parameter name="energyMeterId" type="text" required="true">
```
Isn't it rather `meterId` now ?

mherwege replied:
I intended to keep it for backward compatibility, but it is probably better to just change it. There aren't many users for energy meters, so the impact shouldn't be that big.

lolodomo commented on bundles/org.openhab.binding.nikohomecontrol/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -186,20 +179,159 @@
+			<channel id="energylast" typeId="measurementtime"/>
+		</channels>
+		<config-description>
+			<parameter name="energyMeterId" type="text" required="true">
```
Isn't it rather `meterId` now ?

lolodomo commented on bundles/org.openhab.binding.nikohomecontrol/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -186,20 +179,159 @@
+			<channel id="gaslast" typeId="measurementtime"/>
 		</channels>
 		<config-description>
 			<parameter name="energyMeterId" type="text" required="true">
```
Isn't it rather `meterId` now ?

lolodomo commented on bundles/org.openhab.binding.nikohomecontrol/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -311,9 +443,77 @@
 		<label>@text/channelPowerLabel</label>
 		<description>@text/channelPowerDescription</description>
 		<category>Number</category>
-		<state readOnly="true" pattern="%.0f %unit%"/>
+		<state readOnly="true"/>
```
Why removing the default pattern ? This is a good practice to define one.

mherwege replied:
Added back in.

lolodomo commented on bundles/org.openhab.binding.nikohomecontrol/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -311,9 +443,77 @@
+		<label>@text/channelEnergyLabel</label>
+		<description>@text/channelEnergyDescription</description>
+		<category>Number</category>
+		<state readOnly="true"/>
```
Same on new channel types, a default pattern would be a good thing.

mherwege replied:
Done

lolodomo reviewed (changes requested):
Review part 1 (resources, constants, handler factory and discovery)

lolodomo commented:
This is a breaking change due to changes in `energyMeter` thing type.

lolodomo commented:
nch1 and nhc2 packages: remains to review NikoHomeControlCommunication1 and NikoHomeControlCommunication2 classes

mherwege commented on bundles/org.openhab.binding.nikohomecontrol/src/main/java/org/openhab/binding/nikohomecontrol/internal/NikoHomeControlBindingConstants.java:
```diff
@@ -46,17 +46,33 @@ public class NikoHomeControlBindingConstants {
+                    .collect(Collectors.toSet()));
+    public static final Set<ThingTypeUID> ACCESS_THING_TYPES_UIDS = Collections
+            .unmodifiableSet(Stream.of(THING_TYPE_ACCESS, THING_TYPE_ACCESS_RINGANDCOMEIN).collect(Collectors.toSet()));
+    public static final Set<ThingTypeUID> SUPPORTED_THING_TYPES_UIDS = Stream
+            .concat(BRIDGE_THING_TYPES_UIDS.stream(),
+                    Stream.concat(ACTION_THING_TYPES_UIDS.stream(),
+                            Stream.concat(THERMOSTAT_THING_TYPES_UIDS.stream(),
+                                    Stream.concat(METER_THING_TYPES_UIDS.stream(), ACCESS_THING_TYPES_UIDS.stream()))))
```
It doesn't, but Steam.of with a flatMap may be more concise.

mherwege commented on bundles/org.openhab.binding.nikohomecontrol/src/main/java/org/openhab/binding/nikohomecontrol/internal/NikoHomeControlBindingConstants.java:
```diff
@@ -46,17 +46,33 @@ public class NikoHomeControlBindingConstants {
-    public static final Set<ThingTypeUID> SUPPORTED_THING_TYPES_UIDS = Collections
-            .unmodifiableSet(Stream.of(THING_TYPE_PUSHBUTTON, THING_TYPE_ON_OFF_LIGHT, THING_TYPE_DIMMABLE_LIGHT,
-                    THING_TYPE_BLIND, THING_TYPE_THERMOSTAT, THING_TYPE_ENERGYMETER).collect(Collectors.toSet()));
+    public static final Set<ThingTypeUID> THERMOSTAT_THING_TYPES_UIDS = Collections
+            .unmodifiableSet(Stream.of(THING_TYPE_THERMOSTAT).collect(Collectors.toSet()));
+    public static final Set<ThingTypeUID> METER_THING_TYPES_UIDS = Collections.unmodifiableSet(
+            Stream.of(THING_TYPE_ENERGYMETER_LIVE, THING_TYPE_ENERGYMETER, THING_TYPE_GASMETER, THING_TYPE_WATERMETER)
+                    .collect(Collectors.toSet()));
+    public static final Set<ThingTypeUID> ACCESS_THING_TYPES_UIDS = Collections
+            .unmodifiableSet(Stream.of(THING_TYPE_ACCESS, THING_TYPE_ACCESS_RINGANDCOMEIN).collect(Collectors.toSet()));
```
Thanks

mherwege commented on bundles/org.openhab.binding.nikohomecontrol/src/main/resources/OH-INF/i18n/nikohomecontrol.properties:
```diff
@@ -43,27 +43,43 @@ blindDescription = Rollershutter type action in Niko Home Control
+accessRingAndComeInLabel = Ring And Come In Access Control
+accessRingAndComeInDescription = Access Control with ring and come in function in the Niko Home Control system
+
+deviceConfigDeviceIdLabel = Action ID
+deviceConfigDeviceIdDescription = Niko Home Control action ID
```
Should be device

mherwege commented:
@lolodomo Thank you for your review so far. I believe I have responded to your requests.

lolodomo commented:
> @lolodomo Thank you for your review so far. I believe I have responded to your requests.

This is stil in my TODO list to finish the review (with the 3.4 final release as deadline for that).

lolodomo commented:
Sorry @mherwege , I finally did not yet fully review your PR.

PS: there is a conflict that will need to be fixed.

mherwege commented:
I think I resolved the conflict. That required a force push.

jlaur commented:
@mherwege - FYI, there are now some conflicts to resolve.

lolodomo commented:
> @lolodomo - gentle ping, still on your TODO list to finish this review?

Unfortunately, I will not have enough time before OH4 is released.

mherwege commented:
As much as I would like this to be included, I won’t have time to include review comments before OH4 release myself. I would like to move it forward soon after though. There is too much other things I would like to do with this I can’t really progress without getting this done.

mherwege commented:
@lolodomo @jlaur Rebased. All conflicts should be resolved, ready for review.

mherwege commented:
Rebased

mherwege commented:
@lolodomo You were working on reviewing this over a year ago. I just rebased and would appreciate if we could work towards getting this in 4.2. I actually have code to follow after this one, but was refraining from it as long as this one had not been accepted. Is there any chance you could pick up reviewing this again?

mherwege commented:
@lolodomo I appreciate all your efforts. I am sorry to insist requesting a review of this PR though. This has been waiting for review for a very long time now. Last review activity was almost 2 years ago.

lolodomo commented:
I can't these days but I will continue the review soon.

lolodomo commented on bundles/org.openhab.binding.nikohomecontrol/src/main/java/org/openhab/binding/nikohomecontrol/internal/discovery/NikoHomeControlDiscoveryService.java:
```diff
@@ -78,79 +80,121 @@ public void discoverDevices() {
+                            METER_ID, deviceId, thingName, thingLocation);
+                    break;
+                case GAS:
+                    addDevice(new ThingUID(THING_TYPE_GASMETER, bridgeHandler.getThing().getUID(), deviceId), deviceId,
+                            METER_ID, thingName, thingLocation);
```
I have the feeling you inverted the second and third parameter ?

mherwege replied:
Indeed, thanks

lolodomo commented on bundles/org.openhab.binding.nikohomecontrol/src/main/resources/OH-INF/i18n/nikohomecontrol.properties:
```diff
@@ -112,15 +158,10 @@ offline.configuration-error.ip = Cannot resolve bridge IP with given host name
-offline.configuration-error.energyMeterId = Configured energy meter ID does not match an energy meter in controller
-offline.configuration-error.energyMeterRemoved = Energy meter has been removed from controller
-
-offline.configuration-error.thermostatId = Configured thermostat ID does not match a thermostat in controller
```
This entry is still used in class ThermostantHandler

mherwege replied:
Changed

lolodomo commented on bundles/org.openhab.binding.nikohomecontrol/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -313,7 +446,75 @@
+		<item-type>Number:Energy</item-type>
+		<label>@text/channelEnergyLabel</label>
+		<description>@text/channelEnergyDescription</description>
+		<category>Number</category>
```
I would suggest to not set the category "Number" as the icon is missing in the classic icon set.
Same for all other channel types.

mherwege replied:
Yes, thanks

lolodomo reviewed (changes requested):
Restart of the review

mherwege commented:
@lolodomo I addressed your comments, but had to do a force push after a rebase. There was something wrong with my local environment.

mherwege commented on bundles/org.openhab.binding.nikohomecontrol/src/main/java/org/openhab/binding/nikohomecontrol/internal/protocol/nhc2/NikoHomeControlCommunication2.java:
I have a compiler error in my Eclipse IDE on this file in line 693 and 695 (Null type mismatch (type annotations): required 'Optional<@Nullable String>' but this expression has type '@NonNull Optional<@NonNull String>').
Removing the @Nullable from the declariation solves it for the IDE. But it does not compile on the command line in that case.
These lines where not changed. Any idea what the change in the development setup could be?

lsiepel replied:
I guess the compiler doesn't understand this due to the Optional wrapper. Anyway, what is the use of an Optional<@Nullable String> ? 

If you want to use it you first have to check the optional and later on the string for nullability? Feels kinda strange.

mherwege replied:
> Anyway, what is the use of an Optional<@nullable String>

That's the whole point. It shouldn't be @nullable in the first place (and it actually is not, if you look at the lines it even filters null elements in the stream). It was just put in there in the past to eliminate compiler failures. I didn't touch this code in this PR. Eclipse got more intelligent and gives, in my eyes the correct, warning suggesting to remove the @nullable annotation. But if I do that the build will fail.
This is not breaking and we should of course make sure the build works.

lsiepel replied:
Well, i meant to get rid of all the `Optionals` :-) i have never liked them, but maybe that is a personal preference. 
Becasue the nullable (or non nullable ) string is wrapped in an optional, it seems to be hidden. anyway, as this is not affecting the build, i would either get rid of the optionals or leave it as is.

lsiepel replied:
Don't see to work around it without an extensive rewrite. Adding the @Nullable annotation is perfectly fine for now.

mherwege commented:
@lolodomo I am sorry about insisting. Is there anything I need to do to make this progress? I would like to get this done before this PR’s second anniversary.

lolodomo commented on bundles/org.openhab.binding.nikohomecontrol/src/main/java/org/openhab/binding/nikohomecontrol/internal/protocol/NhcAccess.java:
```diff
@@ -0,0 +1,338 @@
+    /**
+     * Set location of the access control device.
+     *
+     * @param access control location
```
To be fixed: parameter name is "location"

lolodomo commented on bundles/org.openhab.binding.nikohomecontrol/src/main/java/org/openhab/binding/nikohomecontrol/internal/protocol/NhcAccess.java:
```diff
@@ -0,0 +1,338 @@
+    /**
+     * Get state of the access control device.
+     *
+     * @return action state
```
**bell** state

lolodomo commented on bundles/org.openhab.binding.nikohomecontrol/src/main/java/org/openhab/binding/nikohomecontrol/internal/protocol/NhcAccess.java:
```diff
@@ -0,0 +1,338 @@
+    /**
+     * Get state of the access control device.
+     *
+     * @return action state
```
**door lock** state

lolodomo commented on bundles/org.openhab.binding.nikohomecontrol/src/main/java/org/openhab/binding/nikohomecontrol/internal/protocol/NikoHomeControlCommunication.java:
```diff
@@ -166,12 +166,30 @@ public Map<String, NhcThermostat> getThermostats() {
 
     /**
-     * Return all energyMeters meters in the Niko Home Control Controller.
+     * Return all meters meters in the Niko Home Control Controller.
```
one "meters" in double

lolodomo reviewed (changes requested):
Few minor comments for the protocol package, mainly javadoc, except my comment about method visibility in interfaces.

lolodomo commented on bundles/org.openhab.binding.nikohomecontrol/src/main/java/org/openhab/binding/nikohomecontrol/internal/handler/NikoHomeControlActionHandler.java:
```diff
@@ -103,25 +79,24 @@ private void handleCommandSelection(ChannelUID channelUID, Command command) {
                 logger.debug("unexpected command for channel {}", channelUID.getId());
         }
+
+        updateStatus(ThingStatus.ONLINE);
```
Why are you updating the thing status when a command is sent to a channel ? I don't understand the logic.

Same comment for all classes implementing handleCommandSelection

mherwege replied:
Not needed.

lolodomo commented on bundles/org.openhab.binding.nikohomecontrol/src/main/java/org/openhab/binding/nikohomecontrol/internal/handler/NikoHomeControlBaseHandler.java:
```diff
@@ -0,0 +1,140 @@
+        ThingStatus bridgeStatus = bridgeStatusInfo.getStatus();
+        if (ThingStatus.ONLINE.equals(bridgeStatus)) {
+            if (!initialized) {
+                scheduler.submit(this::startCommunication);
```
This thread is not monitored.
I imagine it could be required to "cancel" it in case the thing handler is disposed.

mherwege replied:
Yes, indeed, I guess this could happen when disposing during the start of the communication. I will monitor and cancel in dispose.

lolodomo commented:
Remains to review: README and 4 classes: NikoHomeControlAccessHandler, NikoHomeControlMeterHandler, NikoHomeControlCommunication1 and NikoHomeControlCommunication2

lolodomo commented on bundles/org.openhab.binding.nikohomecontrol/src/main/java/org/openhab/binding/nikohomecontrol/internal/handler/NikoHomeControlMeterHandler.java:
```diff
@@ -0,0 +1,324 @@
+            int value = (invert ? -1 : 1) * power;
+            updateState(CHANNEL_POWER, new QuantityType<>(value, Units.WATT));
+        }
+        updateStatus(ThingStatus.ONLINE);
```
Same remark as in NikoHomeControlAccessHandler

lolodomo commented on bundles/org.openhab.binding.nikohomecontrol/src/main/java/org/openhab/binding/nikohomecontrol/internal/handler/NikoHomeControlMeterHandler.java:
```diff
@@ -0,0 +1,324 @@
+                updateState(CHANNEL_ENERGY, new QuantityType<>(value, KILOWATT_HOUR));
+                updateState(CHANNEL_ENERGY_DAY, new QuantityType<>(dayValue, KILOWATT_HOUR));
+                updateState(CHANNEL_ENERGY_LAST, new DateTimeType(lastReading));
+                updateStatus(ThingStatus.ONLINE);
```
Same

lolodomo commented on bundles/org.openhab.binding.nikohomecontrol/src/main/java/org/openhab/binding/nikohomecontrol/internal/handler/NikoHomeControlMeterHandler.java:
```diff
@@ -0,0 +1,324 @@
+                updateStatus(ThingStatus.ONLINE);
+                break;
+            default:
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
+                        "@text/offline.configuration-error.meterType");
```
Maybe rather something to test in `startCommunication` ?

mherwege replied:
Yes, could indeed be moved over there.

lolodomo commented on bundles/org.openhab.binding.nikohomecontrol/src/main/java/org/openhab/binding/nikohomecontrol/internal/handler/NikoHomeControlMeterHandler.java:
```diff
@@ -0,0 +1,324 @@
+
+                if (nhcComm.communicationActive()) {
+                    nhcComm.startMeterLive(deviceId);
+                    updateStatus(ThingStatus.ONLINE);
```
Same here and in channelUnlinked

lolodomo reviewed (changes requested):
End of my review of the handler package

mherwege commented on bundles/org.openhab.binding.nikohomecontrol/src/main/java/org/openhab/binding/nikohomecontrol/internal/protocol/NhcBaseEvent.java:
```diff
@@ -0,0 +1,39 @@
+     * Called to indicate the device has been initialized.
+     *
+     */
+    public void deviceInitialized();
```
I had seen this, but this PR was prepared before the cleanup PR, so didn't think about it anymore. Will change everywhere.

mherwege commented on bundles/org.openhab.binding.nikohomecontrol/src/main/java/org/openhab/binding/nikohomecontrol/internal/handler/NikoHomeControlAccessHandler.java:
```diff
@@ -0,0 +1,300 @@
+        }
+
+        if (nhcAccess.supportsVideoStream()) {
+            properties.put("username", "admin");
+            properties.put("password", "123qwe");
```
The binding does not in itself support using the video stream from the video door device, but it could be configured using the IP Camera binding. To be able to configure, the username, password and URL's would be required. Putting this in properties is just a convenience if one wants to configure this, but not strictly required for the functioning of the binding. The video device uses a fixed password.

lolodomo replied:
Maybe rather an information to put in the documentation ?

mherwege replied:
I would rather not write it out in the documentation, especially the password.

lolodomo replied:
Ok. If you think there is fine to have this as properties, I close my comment.

mherwege replied:
I think it is fine.

mherwege commented on bundles/org.openhab.binding.nikohomecontrol/src/main/java/org/openhab/binding/nikohomecontrol/internal/handler/NikoHomeControlAccessHandler.java:
```diff
@@ -0,0 +1,300 @@
+            return;
+        }
+
+        updateState(CHANNEL_BELL_BUTTON, state ? OnOffType.ON : OnOffType.OFF);
```
state is a boolean. the OnOffType.valueOf method expects a String as argument.

lolodomo replied:
There is a method, I probably did not give you the good one. The method I am thinking about is taking a boolean as parameter.

lolodomo replied:
Sorry the method is `from`: OnOffType.from(state)

mherwege replied:
Done

mherwege commented on bundles/org.openhab.binding.nikohomecontrol/src/main/java/org/openhab/binding/nikohomecontrol/internal/handler/NikoHomeControlAccessHandler.java:
```diff
@@ -0,0 +1,300 @@
+        }
+
+        updateState(CHANNEL_BELL_BUTTON, state ? OnOffType.ON : OnOffType.OFF);
+        updateStatus(ThingStatus.ONLINE);
```
This can be triggered by an event in the connected system. In a number of cases, the OH thing remained offline, but events started comming in through the communications layer. If this happens, the thing is effectively online and can be switched online.

lolodomo replied:
If you are sure it is necessary, that is fine for me.

mherwege commented on bundles/org.openhab.binding.nikohomecontrol/src/main/java/org/openhab/binding/nikohomecontrol/internal/handler/NikoHomeControlBaseHandler.java:
```diff
@@ -0,0 +1,140 @@
+            nhcBridgeHandler.bridgeOnline();
+        } else {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.BRIDGE_UNINITIALIZED,
+                    "@text/offline.bridge-unitialized");
```
I have added it.

lolodomo commented:
I will try to finish tomorrow. README and 2 last classes. There is no open comments.

lolodomo commented on bundles/org.openhab.binding.nikohomecontrol/src/main/java/org/openhab/binding/nikohomecontrol/internal/protocol/nhc2/NikoHomeControlCommunication2.java:
```diff
@@ -372,94 +377,215 @@ private void addDevice(NhcDevice2 device) {
+                    accessType = AccessType.RINGANDCOMEIN;
+                }
+            }
+            accessType = AccessType.BASE;
```
Suspect: you overwrite the value you potentially set at line 480.

lolodomo commented on bundles/org.openhab.binding.nikohomecontrol/src/main/java/org/openhab/binding/nikohomecontrol/internal/protocol/nhc2/NikoHomeControlCommunication2.java:
```diff
@@ -372,94 +377,215 @@ private void addDevice(NhcDevice2 device) {
+                        int buttonIndex = access.getButtonIndex();
+                        logger.debug("link access device {} to video device {} button {}", device.uuid,
+                                videoDevice.getId(), buttonIndex);
+                        videoDevice.setNhcAccess(access.getButtonIndex(), access);
```
Suggestion: you can use variable `buttonIndex`.

lolodomo commented on bundles/org.openhab.binding.nikohomecontrol/src/main/java/org/openhab/binding/nikohomecontrol/internal/protocol/nhc2/NikoHomeControlCommunication2.java:
```diff
@@ -614,14 +748,66 @@ private void updateEnergyMeterState(NhcEnergyMeter2 energyMeter, List<NhcPropert
         } catch (NumberFormatException e) {
-            energyMeter.setPower(null);
-            logger.trace("wrong format in energy meter {} power reading", energyMeter.getId());
+            logger.trace("wrong format  in energy meter {} power reading", meter.getId());
```
Ultra minor: a double space

lolodomo commented on bundles/org.openhab.binding.nikohomecontrol/src/main/java/org/openhab/binding/nikohomecontrol/internal/protocol/nhc2/NikoHomeControlCommunication2.java:
```diff
@@ -793,27 +984,133 @@ public void startEnergyMeter(String energyMeterId) {
+        NhcMessage2 message = new NhcMessage2();
+
+        message.method = "devices.control";
+        ArrayList<NhcMessageParam> params = new ArrayList<>();
```
`List<NhcMessageParam>`

lolodomo commented on bundles/org.openhab.binding.nikohomecontrol/src/main/java/org/openhab/binding/nikohomecontrol/internal/protocol/nhc2/NikoHomeControlCommunication2.java:
```diff
@@ -793,27 +984,133 @@ public void startEnergyMeter(String energyMeterId) {
+        NhcMessageParam param = new NhcMessageParam();
+        params.add(param);
+        message.params = params;
+        ArrayList<NhcDevice2> devices = new ArrayList<>();
```
`List<NhcDevice2>`

lolodomo commented on bundles/org.openhab.binding.nikohomecontrol/src/main/java/org/openhab/binding/nikohomecontrol/internal/protocol/nhc2/NikoHomeControlCommunication2.java:
```diff
@@ -793,27 +984,133 @@ public void startEnergyMeter(String energyMeterId) {
+        devices.add(device);
+        param.devices = devices;
+        device.uuid = accessId;
+        ArrayList<NhcProperty> deviceProperties = new ArrayList<>();
```
`List<NhcProperty>`

lolodomo commented on bundles/org.openhab.binding.nikohomecontrol/src/main/java/org/openhab/binding/nikohomecontrol/internal/protocol/nhc2/NikoHomeControlCommunication2.java:
```diff
@@ -793,27 +984,133 @@ public void startEnergyMeter(String energyMeterId) {
+        NhcMessage2 message = new NhcMessage2();
+
+        message.method = "devices.control";
+        ArrayList<NhcMessageParam> params = new ArrayList<>();
```
`List<NhcMessageParam>`

lolodomo commented on bundles/org.openhab.binding.nikohomecontrol/src/main/java/org/openhab/binding/nikohomecontrol/internal/protocol/nhc2/NikoHomeControlCommunication2.java:
```diff
@@ -793,27 +984,133 @@ public void startEnergyMeter(String energyMeterId) {
+        NhcMessageParam param = new NhcMessageParam();
+        params.add(param);
+        message.params = params;
+        ArrayList<NhcDevice2> devices = new ArrayList<>();
```
`List<NhcDevice2>`

lolodomo commented on bundles/org.openhab.binding.nikohomecontrol/src/main/java/org/openhab/binding/nikohomecontrol/internal/protocol/nhc2/NikoHomeControlCommunication2.java:
```diff
@@ -793,27 +984,133 @@ public void startEnergyMeter(String energyMeterId) {
+        devices.add(device);
+        param.devices = devices;
+        device.uuid = accessId;
+        ArrayList<NhcProperty> deviceProperties = new ArrayList<>();
```
`List<NhcProperty>`

lolodomo commented on bundles/org.openhab.binding.nikohomecontrol/src/main/java/org/openhab/binding/nikohomecontrol/internal/protocol/nhc2/NikoHomeControlCommunication2.java:
```diff
@@ -793,27 +984,133 @@ public void startEnergyMeter(String energyMeterId) {
+        NhcMessage2 message = new NhcMessage2();
+
+        message.method = "devices.control";
+        ArrayList<NhcMessageParam> params = new ArrayList<>();
```
`List<NhcMessageParam>`

lolodomo commented on bundles/org.openhab.binding.nikohomecontrol/src/main/java/org/openhab/binding/nikohomecontrol/internal/protocol/nhc2/NikoHomeControlCommunication2.java:
```diff
@@ -793,27 +984,133 @@ public void startEnergyMeter(String energyMeterId) {
+        NhcMessageParam param = new NhcMessageParam();
+        params.add(param);
+        message.params = params;
+        ArrayList<NhcDevice2> devices = new ArrayList<>();
```
`List<NhcDevice2>`

lolodomo commented on bundles/org.openhab.binding.nikohomecontrol/src/main/java/org/openhab/binding/nikohomecontrol/internal/protocol/nhc2/NikoHomeControlCommunication2.java:
```diff
@@ -793,27 +984,133 @@ public void startEnergyMeter(String energyMeterId) {
+        devices.add(device);
+        param.devices = devices;
+        device.uuid = accessId;
+        ArrayList<NhcProperty> deviceProperties = new ArrayList<>();
```
`List<NhcProperty>`

lolodomo reviewed (changes requested):
Review of class NikoHomeControlCommunication2

lolodomo reviewed (changes requested):
Review of NikoHomeControlCommunication1 : just one ultra minor comment

lolodomo commented:
Just remain to have a look to README changes.

lolodomo commented on bundles/org.openhab.binding.nikohomecontrol/README.md:
```diff
@@ -24,13 +24,29 @@ The installation only needs to be 'connected' (registered on the Niko Home Contr
+The following thing types are available in the binding:
+
+| Thing Type          | NHC I | NHC II | Description                                                                       |
+|---------------------|:-----:|:------:|-----------------------------------------------------------------------------------|
```
Is it normal to have ":" in your different tables for this second line ?
What is the behaviour when used ?

mherwege replied:
:--: means the content of the column will be cneter aligned. (:-- for left aligned, --: for right aligned).

lolodomo commented on bundles/org.openhab.binding.nikohomecontrol/README.md:
````diff
@@ -161,106 +153,59 @@ Note down the `actionId` parameter from the thing, remove it before adding it ag
 
 ```java
-energymeter <thingId> "Label" @ "Location" [ energyMeterId="<Niko Home Control energy meter ID>" ]
+Thing nikohomecontrol:accessRingAndComeIn:mybridge:myaccess [ accessId="abcdef01-dcba-1234-ab98-012345abcdef", refresh=30 ]
````
refresh parameter is not mentioned as applicable for the accessRingAndComeIn thing type in the table above.

lolodomo commented on bundles/org.openhab.binding.nikohomecontrol/README.md:
```diff
@@ -337,6 +280,8 @@ Selection item=ThermostatMode mappings=[Day="day", Night="night", Eco="eco", Off
 Setpoint item=SetTemperature minValue=0 maxValue=30
 Slider item=OverruleDuration minValue=0 maxValue=120
 Text item=Power
```
CurPower

lolodomo commented on bundles/org.openhab.binding.nikohomecontrol/README.md:
```diff
@@ -161,106 +153,59 @@ Note down the `actionId` parameter from the thing, remove it before adding it ag
+| heatingmode     | RW |          | String             | thermostat  | current thermostat mode. Allowed values are Day, Night, Eco, Off, Cool, Prog1, Prog2, Prog3. Setting `heatingmode` will reset the `setpoint` channel to the standard value for the mode in the controller |
+| mode            | RW |    X     | Number             | thermostat  | current thermostat mode, same meaning as `heatingmode`, but numeric values (0=Day, 1=Night, 2=Eco, 3=Off, 4=Cool, 5=Prog1, 6=Prog2, 7=Prog3). Setting `mode` will reset the `setpoint` channel to the standard value for the mode in the controller. This channel is kept for binding backward compatibility |
+| setpoint        | RW |          | Number:Temperature | thermostat  | current thermostat setpoint. Updating the `setpoint` will overrule the temperature setpoint defined by `heatingmode` or `mode` for `overruletime` duration. Because of API restrictions, NHC II will only report in 0.5°C increments |
+| overruletime    | RW |          | Number             | thermostat  | used to set the total duration to apply the setpoint temperature set in the `setpoint` channel before the thermostat returns to the setting from its mode |
```
Maybe you could mention the unit of this duration.

lolodomo reviewed (changes requested):
Review of README and end of review

mherwege commented:
@lolodomo Many thanks for your thorough review. I believe I have addressed your comments. I will try to do a final smoke test on Friday, if all still works (I am not home at the moment). If you don't have other remarks and that test is OK, I believe this should be good to go.

lolodomo reviewed (approved):
LGTM, thank you

mherwege commented:
@lolodomo I did a quick smoke test in my installation and it looks good for me. So this would be ready as far as I am concerned.

lolodomo commented:
Thank you for your patience @mherwege
